flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/PushBasedAsyncDataInput.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.runtime.io;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.runtime.io.AsyncDataInput;
+import org.apache.flink.runtime.io.AvailabilityListener;
+import org.apache.flink.streaming.api.watermark.Watermark;
+import org.apache.flink.streaming.runtime.streamrecord.LatencyMarker;
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
+import org.apache.flink.streaming.runtime.streamstatus.StreamStatus;
+
+/**
+ * The variant of {@link AsyncDataInput} that is defined for handling both network
+ * input and source input in a unified way via {@link #emitNext(DataOutput)} instead
+ * of returning {@code Optional.empty()} via {@link AsyncDataInput#pollNext()}.
+ */
+@Internal
+public interface PushBasedAsyncDataInput extends AvailabilityListener {
```

Comment:
(loudly thinking) maybe rename `PushBasedAsyncDataInput` -> `PushingAsyncDataInput` and `AsyncDataInput` -> `PollingAsyncDataInput`?